src/main/scala/mesosphere/marathon/core/matcher/base/util/OfferOperationFactory.scala

```diff
@@ -0,0 +1,94 @@
+package mesosphere.marathon.core.matcher.base.util
+
+import mesosphere.marathon.WrongConfigurationException
+import mesosphere.marathon.core.task.Task.LocalVolume
+import mesosphere.marathon.state.PathId
+import org.apache.mesos.Protos.Resource.ReservationInfo
+import org.apache.mesos.{ Protos => Mesos }
+import org.slf4j.LoggerFactory
+
+class OfferOperationFactory(
+    private val principalOpt: Option[String],
+    private val roleOpt: Option[String]) {
+
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  private[this] lazy val principal: String = principalOpt match {
+    case Some(value) => value
+    case _ => throw new WrongConfigurationException(
+      "No principal set. Set --mesos_authentication_principal to enable using local volumes in Marathon.")
+  }
+
+  private[this] lazy val role: String = roleOpt match {
+    case Some(value) => value
+    case _ => throw new WrongConfigurationException(
+      "No principal set. Set --mesos_role to enable using local volumes in Marathon.")
+  }
+
+  /** Create a launch operation for the given taskInfo. */
+  def launch(taskInfo: Mesos.TaskInfo): Mesos.Offer.Operation = {
+    val launch = Mesos.Offer.Operation.Launch.newBuilder()
+      .addTaskInfos(taskInfo)
+      .build()
+
+    Mesos.Offer.Operation.newBuilder()
+      .setType(Mesos.Offer.Operation.Type.LAUNCH)
+      .setLaunch(launch)
+      .build()
+  }
+
+  def reserve(resources: Iterable[Mesos.Resource]): Mesos.Offer.Operation = {
+    import scala.collection.JavaConverters._
+    val reservedResources = resources.map { resource =>
+      Mesos.Resource.newBuilder(resource)
+        .setRole(role) // actually needed or already in the resource?
```

Comment:
Don't understand the question in the comment. Question for the mesos folks, we should answer before merge?